### ISSUE TITLE
Ubuntu 18.04 Support

### ISSUE THREAD

**Issue by ajcollett** [derdanne/puppet-nfs#83]
So I have tried this module out on Ubuntu 18.04. 

I know it does not officially support 18.04, but I thought it would be good to help get 18.04 supported. 

There exists the same issue as in: #57 

i.e. that 18.04 uses the `nfs-common` package instead of `idmapd`.

I'm going to try hack this into the module locally, and see what other issues come up.

**Comment by ajcollett** [comment]
I just added `'bionic'` to the list including '`stretch`' in `params.pp` and it seems to work. Mounted directory and puppet runs happily. Let me know if you'd like me to any more tests for you.

**Comment by derdanne** [comment]
Hi there! @ajcollett!

Thanks! It would be nice if you could give me a PR with your changes.

Best regards

**Comment by barcharcraz** [comment]
having this problem on fedora-28 as well.

**Comment by derdanne** [comment]
Hey, thanks for the PR. I had the chance to look at the Ubuntu 18.04 bionic support. I had to add some more changes in params.pp to get it work. I have also added spec and acceptance tests for bionic.  Please view the changes in develop branch.

**Comment by jadestorm** [comment]
@derdanne I'm still testing against 18, but one thing I caught is that 18 uses a different server service name as well -- instead of nfs-server, it's nfs-kernel-server.  (see https://linuxconfig.org/how-to-configure-a-nfs-file-server-on-ubuntu-18-04-bionic-beaver)

**Comment by jadestorm** [comment]
(note: things aren't working out quite right even after I made that change -- still poking around)

**Comment by jadestorm** [comment]
Ok I'm not going to claim to understand why so many things seem "missing" in Ubuntu 18 -- guessing most of this is incorporated into the kernel now.  Either way, this is what ended up actually working for me:
```

        'bionic': {
          $client_services            = {'rpcbind' => {}}
          $client_nfsv4_services      = {'rpcbind' => {}}
          $server_service_name        = 'nfs-kernel-server'
        }
```

**Comment by derdanne** [comment]
Hey, please have a look at the current master, i've added bionic support including acceptance testing.

**Comment by jadestorm** [comment]
Howdy!  That's what I was testing against I'm afraid.  It was obvious to me that you had tested it already from this thread, and yet it would just not work for me until I made the modification I listed above.  (which included removing the nfs-server part and also the idmap part)  I don't entirely understand where the disconnect might be.  =/  (nfs-server and nfs-idmapd don't exist as services on my install)

lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.1 LTS
Release:        18.04
Codename:       bionic

if that helps anything but I doubt it does.  Any ideas why we might be seeing different things?

**Comment by derdanne** [comment]
Hey, i've used the official Ubuntu bionic docker image in my nodeset (https://github.com/derdanne/puppet-nfs/blob/master/spec/acceptance/nodesets/ubuntu-18.04.yml) to run the acceptance test. As you can see here, the current codebase works on the container:

https://travis-ci.org/derdanne/puppet-nfs/jobs/402245817
`nfs class
  include nfs without params
    when default parameters
localhost $ scp /tmp/beaker20180710-14841-1ovjvow ubuntu-1804-x64:/tmp/apply_manifest.pp.l30PUW {:ignore => }
      applies without errors without params
  include nfs with server params
    when server params
localhost $ scp /tmp/beaker20180710-14841-rc5wwu ubuntu-1804-x64:/tmp/apply_manifest.pp.MAQPmG {:ignore => }
      works with no errors based on the example
localhost $ scp /tmp/beaker20180710-14841-13w8bz4 ubuntu-1804-x64:/tmp/apply_manifest.pp.GriRzv {:ignore => }
      runs a second time without changes
      Package "nfs-common"
        should be installed
      Package "nfs-kernel-server"
        should be installed
      Package "nfs4-acl-tools"
        should be installed
      Package "rpcbind"
        should be installed
      Service "nfs-server"
        should be running
      Service "nfs-idmapd"
        should be running`

Is this a fresh install, or did you upgrade the from 16.04 to 18.04?

**Comment by jadestorm** [comment]
That is -very- strange.  This was a fresh install.  Let me paste some information here but then I might have a theory I can test shortly.

```
root@mesmer-ub18:~# systemctl -a | grep nfs
● nfs-blkmap.service                                                                                                                                        not-found inactive dead      nfs-blkmap.service                                     
  nfs-config.service                                                                                                                                        loaded    inactive dead      Preprocess NFS configuration                           
  nfs-utils.service                                                                                                                                         loaded    inactive dead      NFS server and client services                         
  nfs-client.target                                                                                                                                         loaded    inactive dead      NFS client services             
```                       

```
root@mesmer-ub18:~# dpkg -l | grep nfs
ii  libnfsidmap2:amd64                    0.25-5.1                          amd64        NFS idmapping library
ii  nfs-common                            1:1.3.4-2.1ubuntu5                amd64        NFS support files common to client and server
ii  nfs4-acl-tools                        0.3.3-3                           amd64        Commandline and GUI ACL utilities for the NFSv4 client
```

Now one thing you'll notice I do -not- have.  nfs-kernel-server.  Basically I was aiming to configure this as just a client, not serve anything itself.  Is there a chance that the "client only" config is not being tested?  Or have you already tested that as well?

Here's the params I pass in via hiera:
```
nfs::client_enabled: true
nfs::nfs_v4_client: true
nfs::nfs_v4_idmap_domain: 'XXXXX'
nfs::manage_server_service: false
```

Am I doing something foolish in my config then?

**Comment by jadestorm** [comment]
One additional thing I will add -- this is a brand new install in all ways -- all I -really- did was make it so puppet compiled successfully and didn't throw errors -- I do not know if the config I went with actually has a functional nfs4 client setup on it.  From the variety of articles I read on this last week, it appeared that Ubuntu 18 really shook things up with the way NFS is set up and it was not clear to me if any nfs or idmap daemon would actually be running for a client type setup.

**Comment by kratan** [comment]
ubuntu 18.04. hmm, there is no idmapd for systemd, maybe killing rpc.idmapd if running and starting it would be a solution?

```
Error: Systemd start for idmapd failed!
journalctl log for idmapd:
-- Logs begin at Mon 2018-08-27 16:18:20 CEST, end at Mon 2018-09-10 16:44:06 CEST. --
-- No entries --

Error: /Stage[main]/Nfs::Client::Service/Service[idmapd]/ensure: change from 'stopped' to 'running' failed: Systemd start for idmapd failed!
journalctl log for idmapd:
-- Logs begin at Mon 2018-08-27 16:18:20 CEST, end at Mon 2018-09-10 16:44:06 CEST. --

```

**Comment by derdanne** [comment]
Hi @kratan,

idmapd in bionic is "nfs-idmapd". I have the changes for the bionic support in the master branch - i will create a release, so this will be in 2.0.9.

**Comment by jadestorm** [comment]
To follow up on my previous comments -- nfs-kernel-server needs to be installed for nfs-idmapd to even be available in systemd.  However, a variety of docs out there seem to indicate you only need that if you are running an NFS server rather than running it as a client.  I did confirm that by installing nfs-kernel-server I was able to find nfs-idmapd.  That said, it's unclear to me whether you need to be running the server components to be able to be a nfs4 client.  (I'm afraid I'm a little foggy on the ins and outs of NFSv4)  I do know that trying to use the module "just configured for a client"  (ie I turned off server) fails -- and I made some tweaks to it accordingly.  On the other side of the spectrum, I haven't tested to see if I can connect to an NFSv4 server yet.  What do you think @derdanne ?  I posted my params above where I configure it for a client "only".

**Comment by stephenkenny** [comment]
Hello - just to report my experience: on Debian 8 and 9, and on Ubuntu 18.04, I'm seeing a similar, or related, phenomenon.  My nodes are configured as NFS clients only, yet on every run of the Puppet agent an attempt is made to start the nfs-server service.  Unsurprisingly it fails.  Here are the parameters, which are looked up in Hiera:

```
  nfs::client_enabled:       true
  nfs::server_enabled:      false
  nfs::nfs_v4_client:        true
  nfs::nfs_v4_idmap_domain:  '<our FQDN>'
  nfs::nfs_v4_mount_root:    '/export'
  nfs::client_nfsv4_options: 'tcp,nolock,rsize=32768,wsize=32768,intr,noatime,actimeo=3,acl,_netdev'
  nfs::manage_server_service: false
```
And this is what I see on every run of the Puppet agent:

```
Error: Could not find init script for 'nfs-server'
Error: /Stage[main]/Nfs::Client::Service/Service[nfs-server]/ensure: change from stopped to running failed: Could not find init script for 'nfs-server'
Notice: /Stage[main]/Nfs::Client/Anchor[nfs::client::end]: Dependency Service[nfs-server] has failures: true
Warning: /Stage[main]/Nfs::Client/Anchor[nfs::client::end]: Skipping because of failed dependencies
```
Thanks for any advice or assistance.

**Comment by derdanne** [comment]
Hi, sorry for the lack of activity here. I will try to work on this in the next week. But i totally agree, that we need to have a look on what is really needed for client-only setups.

**Comment by derdanne** [comment]
Hey @jadestorm, i think i've made some progress on this issue. I've minimalized the service dependencies for ubuntu bionic, like you have suggested. I've reflected these changes in the acceptance tests, and also ran successfully some manual tests on a Ubuntu Bionic Vagrant machine.

@stephenkenny 
I moved back to use the service "nfs-kernel-server" which also works on ubuntu and should work on debian.

For now, the current progress is in the develop branch, when testing is finished, i would merge this into master.

**Comment by jadestorm** [comment]
@derdanne Just wanted to confirm before I test -- are you saying that I should be able to cut over to master right now and have good results?  (note: the Bionic install I have is not production so I"m not worried about breaking anything)  Just wanted to make sure it was good timing to test.  =)

**Comment by derdanne** [comment]
Hey, yes i've just merged the changes into the master branch, you can give it a try :)

**Comment by jadestorm** [comment]
@derdanne Looks great from my testing!  No problems what-so-ever!  =)

**Comment by derdanne** [comment]
Great! Thanks for your support, i've just released 2.0.10 including these changes.
